pkg/core/klinedriver/klinebuilder.go

```diff
@@ -0,0 +1,295 @@
+		lastBucket := *accBucket // make a copy
+		kb.resetKLine(interval, nextBucket.StartTime)
+		// fill the gap with empty trades
+		fillingKLines := lastBucket.FillGapKLines(numShifts)
+		kb.klinesBuffer[interval] = append(kb.klinesBuffer[interval], fillingKLines...)
+
+		kb.lastBucketMap[interval] = &lastBucket
+		accBucket.accumulateTrade(&trade)
+	}
+}
+
+// Update updates the KLineBuilder with the given update time.
+// The semantics of the Update is to notify the current timestamp to the KLineBuilder.
+func (kb *KLineBuilder) Update(updateTime types.Time) (kLinesMap map[types.Interval][]*types.KLine) {
+	kLinesMap = make(map[types.Interval][]*types.KLine)
+	for interval, bufferedKLines := range kb.klinesBuffer {
+		for _, klineRef := range bufferedKLines {
+			kline := *klineRef
+			kLinesMap[interval] = append(kLinesMap[interval], &kline)
+		}
+		kb.klinesBuffer[interval] = make([]*types.KLine, 0)
+	}
+	for interval, accBucket := range kb.accBucketMap {
+		if updateTime.Before(accBucket.StartTime.Time()) {
+			// update time is before the start time, just add kline
+			kline := *accBucket.KLine
+			kLinesMap[interval] = append(kLinesMap[interval], &kline)
+			continue
+		}
+
+		// check if kline is available: there were trades or we have the last kline
+		lastBucket, hasLast := kb.lastBucketMap[interval]
+		if accBucket.KLine.NumberOfTrades == 0 && !hasLast {
+			log.Debugf("kline not available for interval %s(%s) yet", kb.symbol, interval)
+			continue
+		}
+
+		// At this point, either there were trades or we have the last kline
+		if accBucket.KLine.NumberOfTrades == 0 {
+			// no trade happened up to the updateTime
+			// use the last kline to fill the gap
+			accBucket.KLine.Open = lastBucket.KLine.Close
+			accBucket.KLine.High = lastBucket.KLine.Close
+			accBucket.KLine.Low = lastBucket.KLine.Close
+			accBucket.KLine.Close = lastBucket.KLine.Close
+			accBucket.KLine.LastTradeID = lastBucket.KLine.LastTradeID
+			accBucket.KLine.Volume = fixedpoint.Zero
+			accBucket.KLine.QuoteVolume = fixedpoint.Zero
+			accBucket.KLine.TakerBuyBaseAssetVolume = fixedpoint.Zero
+			accBucket.KLine.TakerBuyQuoteAssetVolume = fixedpoint.Zero
+		}
+		// two cases from here:
+		// - the update time is in the accumulating bucket
+		// - the update time is after the accumulating bucket
+		//   - there might be multiple buckets in between
+
+		// check if the update time is in the accumulating bucket
+		if accBucket.Contains(updateTime) {
+			// update time is in the accumulating bucket, just add kline
+			accBucket.KLine.EndTime = types.Time(updateTime.Time().UTC())
+			kline := *accBucket.KLine
+			kLinesMap[interval] = append(kLinesMap[interval], &kline)
+			continue
+		}
+
+		// update time is after the accumulating bucket
+		// we should close the current accumulating bucket and add kline
+		accBucket.KLine.Closed = true
+		accBucket.KLine.EndTime = types.Time(accBucket.EndTime.Time().UTC())
```

Comment:
no -1ms here?

Reply:
I prefer to make the logic in `KLineBuilder` as simple as possible.
It's easier to test and also easier to control its behaviour.
Any adjustment to the start/end time will be done in the driver.

Reply:
ok, then it would be better to have a comment about this above this line, explain why it's not EndTime - 1ms?

Reply:
ok.